examples/cpp/keyvaluestore/caching_interceptor.h

```diff
@@ -0,0 +1,128 @@
+  void Intercept(
+      ::grpc::experimental::InterceptorBatchMethods* methods) override {
+    bool hijack = false;
+    if (methods->QueryInterceptionHookPoint(
+            grpc::experimental::InterceptionHookPoints::
+                PRE_SEND_INITIAL_METADATA)) {
+      // Hijack all calls
+      hijack = true;
+      // Create a stream on which this interceptor can make requests
+      stub_ = keyvaluestore::KeyValueStore::NewStub(
+          methods->GetInterceptedChannel());
+      stream_ = stub_->GetValues(&context_);
+    }
+    if (methods->QueryInterceptionHookPoint(
+            grpc::experimental::InterceptionHookPoints::PRE_SEND_MESSAGE)) {
+      // We know that clients perform a Read and a Write in a loop, so we don't
+      // need to maintain a list of the responses.
+      std::string requested_key;
+      const keyvaluestore::Request* req_msg =
+          static_cast<const keyvaluestore::Request*>(methods->GetSendMessage());
+      if (req_msg != nullptr) {
+        requested_key = req_msg->key();
+      } else {
+        // The non-serialized form would not be available in certain scenarios,
+        // so add a fallback
+        keyvaluestore::Request req_msg;
+        auto* buffer = methods->GetSerializedSendMessage();
+        auto copied_buffer = *buffer;
+        GPR_ASSERT(
+            grpc::SerializationTraits<keyvaluestore::Request>::Deserialize(
+                &copied_buffer, &req_msg)
+                .ok());
+        requested_key = req_msg.key();
+      }
+
+      // Check if the key is present in the map
+      auto search = cached_map_.find(requested_key);
+      if (search != cached_map_.end()) {
+        std::cout << "Key " << requested_key << "found in map";
+        response_ = search->second;
+      } else {
+        std::cout << "Key " << requested_key << "not found in cache";
+        // Key was not found in the cache, so make a request
+        keyvaluestore::Request req;
+        req.set_key(requested_key);
+        stream_->Write(req);
+        keyvaluestore::Response resp;
+        stream_->Read(&resp);
+        response_ = resp.value();
+        // Insert the pair in the cache for future requests
+        cached_map_.insert({requested_key, response_});
+      }
+    }
+    if (methods->QueryInterceptionHookPoint(
+            grpc::experimental::InterceptionHookPoints::PRE_SEND_CLOSE)) {
+      stream_->WritesDone();
+    }
+    if (methods->QueryInterceptionHookPoint(
+            grpc::experimental::InterceptionHookPoints::PRE_RECV_MESSAGE)) {
+      keyvaluestore::Response* resp =
+          static_cast<keyvaluestore::Response*>(methods->GetRecvMessage());
+      resp->set_value(response_);
+    }
+    if (methods->QueryInterceptionHookPoint(
+            grpc::experimental::InterceptionHookPoints::PRE_RECV_STATUS)) {
+      auto* status = methods->GetRecvStatus();
+      *status = grpc::Status::OK;
+    }
+    if (hijack) {
+      methods->Hijack();
+    } else {
+      methods->Proceed();
```

Comment:
Could you explain here why the Proceed is needed?

Reply:
Yeah, how can a hijacked rpc ever proceed? What does that mean? Isn't PRE_SEND_INITIAL_METADATA the first hook so we're always hijacked?

This needs at least a comment.

Reply:
Proceed indicates that the interceptor is done doing whatever it wanted to do at these hooks, and grpc can decide the next flow of operations .

Reply:
I still have the same question as sheenaqotj

Reply:
Can you add this: "one of Hijack or Proceed always needs to be called to make progress, and Hijack is only to be called once on PRE_SEND_INITIAL_METADATA." I think that really clarified it for me.

Reply:
Cool! Thanks for the feedback! :)